test/SlowTests/Voron/CompactTrees.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using FastTests.Voron;
+using Tests.Infrastructure;
+using Voron.Data.CompactTrees;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace SlowTests.Voron
+{
+
+    public class CompactTreesTest : StorageTest
+    {
+        public CompactTreesTest(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [RavenTheory(RavenTestCategory.Voron)]
+        [InlineData(1337, true)]
+        [InlineData(1337, false)]
+        public void CanRecompressItemsWithDeletesAndInserts(int seed, bool doSampling)
+        {
+            static void Shuffle(string[] list, Random rng)
+            {
+                int n = list.Length;
+                while (n > 1)
+                {
+                    n--;
+                    int k = rng.Next(n + 1);
+                    var value = list[k];
+                    list[k] = list[n];
+                    list[n] = value;
+                }
+            }
+
+            const int Size = 200000;
+
+            Random random = new Random(seed);
+
+            var uniqueKeys = new HashSet<string>();
+            var inTreeKeys = new HashSet<string>();
+            var removedKeys = new HashSet<string>();
+
+            for (int iter = 0; iter < 4; iter++)
+            {
+                using (var wtx = Env.WriteTransaction())
+                {
+                    var tree = CompactTree.Create(wtx.LowLevelTransaction, "test");
+                    for (int i = 0; i < Size; i++)
+                    {
+                        var rname = random.Next();
+                        var key = "hi" + rname;
+                        if (!uniqueKeys.Contains(key))
+                        {
+                            uniqueKeys.Add(key);
+                            inTreeKeys.Add(key);
+                            tree.Add(key, rname);
+                        }
+                    }
+
+                    if (doSampling)
+                        tree.TryImproveDictionaryByRandomlyScanning((int)tree.State.NextTrainAt / 4);
```

Comment:
This does _random_ I/O, which can be quite expensive if we have a large tree, no?

Reply:
A better option would be to compute the number of entries in the tree, and scan that sequentially with random hops through the items in it, that would at least give us better scanning ability.

Reply:
Yes, but not for the performance of the actual dictionary. The closest to random, the better the dictionary because doing say per pages introduce much higher bias. So far we need to access less than 25% of the size of the tree to get a pretty accurate dictionary. There are alternatives, for example a partial sequential access using the knowledge that the keys distribution in `depth - n` layers which is the next one that I will implement on monday to test how effective it can be. So far random makes sense after 10M elements (give or take) because you need just 25% of the tree.

Reply:
Yes, a better option would be:

* Random till you get to a leaf page
* Return all the results from that page

Will need to remember to avoid duplicates, but easy enough to do

Reply:
Not really. Because if you do that you are overrepresenting the clustered nature of leaf nodes and biasing the sample.

Reply:
Current implementation seems to be working pretty good so far. It will sample only 1 key from leaf pages but will get the entire key set from any branch key. So far, it achieves compression numbers comparable with full scanning, with 1/10th of the keys sampled on the test we have. 

Bootstrap dictionary: 35.12 Mb
Full scan: 7.80 Mb
Random branch scan (1/10th of the tree): 7.814 Mb
Random leaf scan (1/10th of the tree):  7.812 Mb

I would call this a done deal... Good part of this, we can actually allow the administrator to run this with different samplings and find rates they want and request a compress on that sampling rate.

Reply:
There is no administrator that can do that. We don't want to go down that road.

In your tests, how many terms do you have there?

Reply:
Synthetic and 100.000 items but they have common prefixes. We need to test with something extracted from a real environment to be sure. What I meant is that the administrator can request a recompress and tell us the expectation: `min`, `normal`, `max` and we can do the sampling to find the diminishing returns point.

Reply:
Remember that this is on a per _term_ basis.
`Synthetic and common prefix` is the most common scenario for `id()` - for example.
I think that this is good enough